### PR TITLE
Adding benchmark to integrator and fixing memory leak.

### PR DESCRIPTION
**Pull request by panmari**
Hitrecord was previously escaping deletion in certain paths. Performance is not impacted from the additional delete calls.

### CODE REVIEW

**Comment by simplay** [comment]
I checked out your branch and and got a segmentation fault when running `./main.x 600 600 6 1`

## Details

This segfault only occurs when rendering images with resolutions >= 600 pixels (at least on my machine).

```bash
~/repos/cpptracer/build (a94c26a) ✔ cmake .. -DCMAKE_BUILD_TYPE=Debug && make
➜  ~/repos/cpptracer/build (a94c26a) ✔ ./main.x 600 600 6 1
C++ Raytracer v1.0
Using 12 threads
Rendering scene Nr 6
Rendering 1 sample(s) per pixel
* * * * free(): invalid pointer
[1]    13646 abort (core dumped)  ./main.x 600 600 6 
```

**Comment by simplay** [comment]
**Update**: commenting out 

```
   93   if (hitRecord->material->hasSpecularRefraction() && ray.depth < MAX_DEPTH) {                                                                                                                                                                                                        
   94   ¦ auto sample = std::unique_ptr<ShadingSample>(                                                                                                                                                                                                                                     
   95   ¦ ¦ ¦ hitRecord->material->evaluateSpecularRefraction(hitRecord.get()));                                                                                                                                                                                                            
   96   ¦ if (sample->isValid) {                                                                                                                                                                                                                                                            
   97   ¦ ¦ refraction.add(*sample->brdf);                                                                                                                                                                                                                                                  
   98                                                                                                                                                                                                                                                                                       
   99   ¦ ¦ const Ray refractedRay(new Vector3f(*hitRecord->position), sample->w, ray.depth + 1);                                                                                                                                                                                           
  100   ¦ ¦ Spectrum* spec = integrate(refractedRay);                                                                                                                                                                                                                                       
  101   ¦ ¦ refraction.mult(*spec);                                                                                                                                                                                                                                                         
  102                                                                                                                                                                                                                                                                                       
  103   ¦ ¦ delete refractedRay.origin;                                                                                                                                                                                                                                                     
  104   ¦ ¦ delete spec;                                                                                                                                                                                                                                                                    
  105   ¦ }                                                                                                                                                                                                                                                                                 
  106   }
```

In `whittedIntegrator.cpp` prevents us from getting the segfault.

File: src/integrators/whittedIntegrator.cpp [hunk 1]
```diff
@@ -88,11 +88,11 @@ Spectrum* WhittedIntegrator::integrate(const Ray& ray) {
       delete reflectedRay.origin;
       delete spec;
     }
-    delete sample;
   }
 
   if (hitRecord->material->hasSpecularRefraction() && ray.depth < MAX_DEPTH) {
-    ShadingSample* sample = hitRecord->material->evaluateSpecularRefraction(hitRecord);
+    auto sample = std::unique_ptr<ShadingSample>(
```
Comment (simplay) [inline src/integrators/whittedIntegrator.cpp]: reverting this change fixes the segmentation fault.

File: src/integrators/whittedIntegrator.cpp [hunk 2]
```diff
@@ -88,11 +88,11 @@ Spectrum* WhittedIntegrator::integrate(const Ray& ray) {
       delete reflectedRay.origin;
       delete spec;
     }
-    delete sample;
```
Comment (simplay) [inline src/integrators/whittedIntegrator.cpp]: in case we revert the change, we should move this `delete sample` into the sample->isValid if-block.

**Comment by simplay** [comment]
after changing ShadingSample's constructor to

```
ShadingSample::~ShadingSample() {
  if (isValid) {
    delete brdf;
    delete emission;
    delete w;
  }
}
```

this PR LGTM - thanks a lot :smile_cat: :+1:

**Comment by panmari** [comment]
I'll have to study the ownership relationship a bit better then. Currently it's rather confusing which class owns which pointer. Maybe the whole thing is even moot, as copying 3 floats might be faster than dealing with ownership relations here.

WDYT, what's the long term direction? Should things like HitRecord, ShadingSample etc. own a copy of their members or do we strive to minimize copies?

**Comment by simplay** [comment]
The final goal is to implement a pathtracer that has an acceptable runtime and memory consumption.

IMO, we should try to implement an intuitive and predictable API. The same time, we also should care about performance (who doesn't like having a raytracer that runs fast? XD).

In other words: Checks like the one I suggested in the ShadingSample's destructor are not intuitive and should be avoided if it doesn't hit performance super-badly. We'd better ensure that all members listed and deleted in a destructor are present and therefore can be deleted.

So, maybe we should ensure that all class-members are initialized after calling a ShadingSample's constructor.

```cpp
// in shadingSample.cpp
ShadingSample::ShadingSample()
  : brdf(new Spectrum), emission(new Spectrum), w(new Vector3f()), isSpecular(false), isValid(false), p(0) {}
```

**Comment by panmari** [comment]
Abandoning in favor of #25
